### ISSUE TITLE
bcftools filter using "." for missing values in integer or float fields

### ISSUE THREAD

**Issue by simonohanlon101** [samtools/bcftools#407]
In `bcftools 1.3` I cannot filter using `"."` to search for missing values in `FORMAT` fields which are defined as `Integer` or `Float`. According to the man page one can filter for missing values and the following examples are given:

> "." to test missing values
>DP=".", DP!=".", ALT="."

This would be useful, e.g. when merging VCF files. For instance, in VCF file [a] we excluded positions which were called as reference, keeping only high quality SNPs, and any positions where we had zero coverage, or where the confidence in the reference call was also low (as we might not want to assume a reference genotype). In VCF file [b] we do the same. Now for positions where [a] does not have a record but [b] does we might end up with a `FORMAT` field which looks like this (for 2 samples and 3 positions):

`1    GT:AD:DP:RGQ    ./.:.:.:.       ./.:0:0:0`
`2    GT:AD:DP:RGQ    ./.:.:.:.       ./.:0:0:0`
`3    GT:AD:DP:RGQ    ./.:.:.:.       ./.:0:0:0`

The first format field is `./.:.:.:.` because it was missing from VCF [a], so we know it was a `0/0` genotype. The second format field is `./.:0:0:0`. The additional information in the `DP` tag tells us that this position had zero coverage so it is good that the genotype is missing.

It would therefore be good to be able to filter these to set the genotype of `./.:.:.:.` back into reference. We *should* be able to achieve this by the following filter command:

`bcftools filter -S0 -e 'FMT/GT="." & FMT/DP = "."' merged.vcf`

However bcftools will not look for missing values using the "." in Integer or Float fields. If I manually edit the header from:

>##FORMAT=<ID=DP,Number=1,Type=Integer,Description="<SNIP>")">

to:

>>##FORMAT=<ID=DP,Number=1,Type=String,Description="<SNIP>")">

The filtering will work as expected and the above filter command returns:

`1    GT:AD:DP:RGQ    0/0:.:.:.       ./.:0:0:0`
`2    GT:AD:DP:RGQ    0/0:.:.:.       ./.:0:0:0`
`3    GT:AD:DP:RGQ    0/0:.:.:.       ./.:0:0:0`

**Comment by simonohanlon101** [comment]
Is this being looked at or considered?

Many thanks,

Simon

**Comment by pd3** [comment]
Yes, this is a known issue on our longer term todo list. It is tied to things like https://github.com/samtools/bcftools/issues/375 and more.

**Comment by simonohanlon101** [comment]
@pd3 many thanks for the reassurances! Cheers and keep up the great work. I LOVE bcftools.
